### ISSUE TITLE
mini.bracketed: treesitter movements for ]t broken and ]T,[T behavior not helpful without working C-i,C-o

### ISSUE THREAD

**Issue by matu3ba** [echasnovski/mini.nvim#297]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)
- [X] I have updated 'mini.nvim' to latest version

### Module(s)

mini.bracketed

### Description

Man page says
```
                                                    *MiniBracketed.treesitter()*
                `MiniBracketed.treesitter`({direction}, {opts})
Tree-sitter node

Go to end/start of current tree-sitter node and its parents (except root).
```
1. Unfortunately this does not clarify what "first, go backwards, go forward and go last" means, because a tree can be walked forwards backwards, but also upwards and downwards.
2. ~~Only `]T` and `[T` appear to work reliably and only for global file scope.~~
3. c++ movements dont work at all, but are likely not focus as one can follow along the brackets for navigation (except when strings contain brackets `{` and `}`.

### Neovim version

NVIM v0.10.0-dev-64+g37bb40701

### Steps to reproduce

1. `nvim -nu minimal.lua` including nvim-treesitter with parsers for python and c++.
2. Try both example files with ]t,[t,]T,[T and compare with ]m, [m, ]M, [M:
```python
class c1:
  def t1(self):
    print(1)
    print(1)
    print(1)
    return 1
    def nested_t1():
      print(1)
      print(1)
      return 1

class c2:
  def t2(self):
    print(2)
    print(2)
    return 2

    def nested_t2():
      print(2)
      return 2
  def t3(self):
    print(2)
    print(2)
    return 2
    def nested_t3():
      print(2)
      return 2

class c3:
  def t2(self):
    print(2)
    print(2)
    return 2
```

For completeness (not discussed):
```cpp
struct s1 {
    struct t1 {
        int testfn1() {
            return 0;
        }
        int a1;
        int a2;
    };

    struct t2 {
        int testfn1() {
            return 0;
            return 0;
            return 0;
        }

        int a1;
        int a2;
    };
};

struct s2 {
    struct t1 {
        int testfn2() {
            return 0;
        }

        int a1;
        int a2;
    };

};


int main() {
    return 0;
}
```

### Expected behavior

1. From `class c1` `]t` as well as `]T` unconditionally jump to last and innermost scope, but only `]T` should do that.
`]t` should jump to end of function.

2. From end of innermost scope (`     return 1` in nested_t1()`), `[t` has expectable behavior, but feels inconsistent as its no reverse operation to `]t` with first jumping to start of statement and then to first statement of block before jumping to end of previous block etc.

3. `[T` works.

4. The other big problem is that `C-i` and `C-o` do not work. Try on a scope `[T` and `C-o`. I would expect that for big movements intermediate jump points would be stored, since they also work from in the mid of class objects.

5. (Personal opinion) Since huge blocks are very common in python classes (but also other languages), I find the behavior for `[T` and `]T` not helpful and instead "moving one scope up or down" sounds like the better alternative (although inconsistent with the other movements).

### Actual behavior

In Summary: 1. `]t`, 2. C-i, 3. C-o, should work 4. `[T`,`]T` is not very useful

**Comment by echasnovski** [comment]
> Unfortunately this does not clarify what "first, go backwards, go forward and go last" means, because a tree can be walked forwards backwards, but also upwards and downwards.

There is also [text after that](https://github.com/echasnovski/mini.nvim/blob/d42d1e5b5394f1a34763dc630f70dffe2470ef84/doc/mini-bracketed.txt#L417-L427):
- > Direction "forward" moves cursor forward to node's end, "backward" - backward to node's start.
- > Directions "first" and "last" work differently from most other targets for performance reasons. They are essentially "backward" and "forward" with very big `n_times` option.

------

As mentioned in the help file, 'treesitter' target moves based on tree created with tree-sitter. Logic is as follows:
- Get smallest node under cursor.
- Move to its start (`[t`) or end (`]t`).
- If cursor didn't move or current `n_times` is bigger than 1, move to start/end of its direct parent.
- Repeat previous two steps.

As you are on Neovim>=0.9, I highly recommend calling `:InspectTree` and exploring behavior of `]t` along it.

>     1. From `class c1` `]t` as well as `]T` unconditionally jump to last and innermost scope, but only `]T` should do that.
>        `]t` should jump to end of function.

Yes, because when cursor is on `class`, the smallest node is the whole class. Moving to its end means moving towards the last character inside its definition (which happens to be end of inner-most `def` in your example).

> 2\. From end of innermost scope (`     return 1` in `nested_t1()`), `[t`has expectable behavior, but feels inconsistent as its no reverse operation to`]t` with first jumping to start of statement and then to first statement of block before jumping to end of previous block etc.

I don't entirely understand what you mean here, but pressing `[t` from `return 1` in `nested_t1()` moves to start of parent's node because current node is an integer 1 itself (thus moving to its start doesn't move cursor).

> 4 The other big problem is that `C-i` and `C-o` do not work. Try on a scope `[T` and `C-o`. I would expect that for big movements intermediate jump points would be stored, since they also work from in the mid of class objects.

Yeah, adding an option whether previous position should be added to jumplist was on the table but I forgot to add it during beta-testing. I've create #298 for that. Thanks!


>     5. (Personal opinion) Since huge blocks are very common in python classes (but also other languages), I find the behavior for `[T` and `]T` not helpful and instead "moving one scope up or down" sounds like the better alternative (although inconsistent with the other movements).

Consistency here is more important in my opinion.
 
> In Summary: 1. `]t`, 2. C-i, 3. C-o, should work 4. `[T`,`]T` is not very useful

In Summary: 1. `]t` works as expected (if not - that is to blame author of tree-sitter grammar), 2-3. Created #298, 4. That is subject and because of consistency.

------

Closing this as most of the points seem to be expected and there is #298 to track adding to jumplist. If you disagree, feel free to add to this issue.

**Comment by echasnovski** [comment]
@matu3ba, I've pushed new feature to 'mini.bracketed' which adds `add_to_jumplist` to relevant targets. Now you can set it inside `require()` call and `<C-o>` should now work (keep in mind that it only works if jump was on different line):

```lua
require('mini.bracketed').setup({
  treesitter = { options = { add_to_jumplist = true } },
})
```

**Comment by matu3ba** [comment]
Cool, thanks alot!
